src/main/java/relations/r1cs/FileToR1CS.java

```diff
@@ -0,0 +1,478 @@
+//
+//
+//                        }
+//                        return A.iterator();
+//                    });
+//
+//
+//            for (int currRow = 0; currRow < numConstraints; currRow++){
+//                //  Start a fresh row for each of A, B, C
+//                Tuple2<LinearCombination<FieldT>, BufferedReader> resA = makeRowAt(currRow, brA);
+//                final LinearCombination<FieldT> A = resA._1();
+//                brA = resA._2();
+//
+//                Tuple2<LinearCombination<FieldT>, BufferedReader> resB = makeRowAt(currRow, brB);
+//                final LinearCombination<FieldT> B = resB._1();
+//                brB = resB._2();
+//
+//                Tuple2<LinearCombination<FieldT>, BufferedReader> resC = makeRowAt(currRow, brC);
+//                final LinearCombination<FieldT> C = resC._1();
+//                brC = resC._2();
+//
+//                constraints.add(new R1CSConstraint<>(A, B, C));
+//            }
+//
+//            brA.close();
+//            brB.close();
+//            brC.close();
+//
+//        } catch (Exception e){
+//            System.err.println("Error: " + e.getMessage());
+//        }
+//
+//        linearCombinationA.count();
+//        linearCombinationB.count();
+//        linearCombinationC.count();
+//
+//        final R1CSConstraintsRDD<FieldT> constraints = new R1CSConstraintsRDD<>(
+//                linearCombinationA,
+//                linearCombinationB,
+//                linearCombinationC,
+//                numConstraints);
+//
+//        return new R1CSRelationRDD<>(constraints, numInputs, numAuxiliary);
+//    }
+
+    private static <FieldT extends AbstractFieldElementExpanded<FieldT>>
+        LinearCombination<FieldT>
+    JSONObjectToCombination (final JSONObject matrixRow) {
+
+        final LinearCombination<FieldT> L = new LinearCombination<>();
+
+        Iterator<String> keys = matrixRow.keySet().iterator();
+        while (keys.hasNext()) {
+            String key = keys.next();
+            BN254aFields.BN254aFr value;
+            try{
+                value = new BN254aFields.BN254aFr((String) matrixRow.get(key));
+            } catch (ClassCastException e){
+                // Handle case when key-value pairs are String: Long
+                value = new BN254aFields.BN254aFr(Long.toString((long) matrixRow.get(key)));
+            }
+            L.add(new LinearTerm<>(Long.parseLong(key), (FieldT) value));
+        }
+        return L;
+    }
+
+    private static <FieldT extends AbstractFieldElementExpanded<FieldT>>
+        Tuple2<LinearCombination<FieldT>, BufferedReader>
+    makeRowAt (int index, BufferedReader reader) {
+        // Assumes input to be ordered by row and that the last line is blank.
+        final LinearCombination<FieldT> L = new LinearCombination<>();
+
+        try {
+            String nextLine;
+            reader.mark(1);  // TODO - this is very strange.
```

Comment:
https://docs.oracle.com/javase/7/docs/api/java/io/BufferedReader.html#mark(int)

I think this has to be larger than 1, since it is the size in bytes you can look ahead and still reset to the mark. Should probably be the maximum length of a line in our file.

Reply:
Yes. If you understand how these work. I am super uneasy about this at the moment. Even after having read the doc. I've put this on the agenda for Monday.